doc/developer/design/20230528_managed_clusters.md

```diff
@@ -0,0 +1,349 @@
+difficult and has many unknowns, so we intentionally chose to not work on this
+in the past year.
+
+The need for improvement has recently become more urgent, as the designers of
+both our [Terraform provider] and our [web console] have considered layering on
+a better cluster management experience in their respective tools. But the right
+place to solve this is in the database itself, so that we don't need to write
+bespoke logic in each tool that is downstream of the SQL API, and so that users
+who use the SQL API directly get the improved experience too.
+
+So, this design document proposes a new course: a new cluster management API
+that solves most of the ergonomic issues with the existing API but does not
+require the full generality of dynamic cluster scheduling.
+
+## Goals
+
+* Users can manage a cluster's size and replication factor without thinking
+  about individual replicas.
+* We can implement the design in weeks, not quarters.
+
+## Non-Goals
+
+* Automatically resizing clusters without downtime.
+* Automatically resizing clusters in response to load.
+* Automatically rebalancing replicas across availability zones.
+
+## Overview
+
+We'll add support for the following declarative cluster management commands to
+Materialize:
+
+```sql
+-- Create a managed cluster with one small replica.
+CREATE CLUSTER foo SIZE 'small';
+
+-- Size up the cluster. IN THE FUTURE this will gracefully
+-- change the size without downtime. AT PRESENT, this will cause
+-- downtime for as long as it takes the new replica(s) to rehydrate.
+--
+-- If you NEED graceful resizing, convert this cluster to an unmanaged
+-- cluster and handle dropping/creating replicas yourself.
+ALTER CLUSTER foo SIZE 'medium';
+
+-- Add two new replicas automatically.
+ALTER CLUSTER foo REPLICATION FACTOR 3;
+
+-- Turn off the cluster for the night.
+ALTER CLUSTER foo REPLICATION FACTOR 0;
+```
+
+## Detailed description
+
+### Managed clusters
+
+We'll introduce the concept of a **managed** cluster. A managed cluster is one
+with a declared size and replication factor, where Materialize is responsible
+for ensuring the replica set matches the declared size and replication factor.
+The replicas of a managed cluster are visible in the system catalog, but cannot
+be directly modified by users.
+
+Existing clusters, where users control the replica set manually via
+`{CREATE|DROP} CLUSTER REPLICA`, will be deemed **unmanaged** clusters.
+
+### `CREATE CLUSTER`
+
+The `CREATE CLUSTER` statement will learn three new options:
+
+  * `SIZE`, a string, which specifies the desired size of the replicas in the
+    cluster.
+  * `MANAGED`, a boolean, which specifies whether the cluster is managed or
+    unmanaged.
+  * `REPLICATION FACTOR`, an integer, which specifies the desired number of
+    replicas.
+
+Users must specify either the `SIZE` or `REPLICAS` option when creating a
+cluster. The two options may not be specified simultaneously.
```

Comment:
Why can't users specify both the replica size and number of replicas in a single statement?

Reply:
You can specify `SIZE` and `REPLICATION FACTOR` in the same statement! What you can't do is specify `SIZE` and the existing manual replica set configuration with `REPLICAS`:

```
CREATE CLUSTER foo
    SIZE 'small',
    REPLICAS (r1 (SIZE 'medium'), r2 (SIZE 'small'))
```

Reply:
Added an example to demonstrate using `SIZE` and `REPLICATION FACTOR` simultaneously.